indico/web/client/js/custom_elements/ind_datepicker.js

```diff
@@ -0,0 +1,631 @@
+            $calendarButton.removeAttribute('aria-selected');
+          }
+        }
+
+        const $focusableButton =
+          listbox.querySelector('[aria-selected]') || listbox.firstElementChild;
+        $focusableButton.tabIndex = 0;
+      }
+    }
+
+    disconnectedCallback() {
+      this.cleanup();
+    }
+
+    static observedAttributes = ['open', 'value', 'range-start', 'range-end'];
+
+    attributeChangedCallback(name) {
+      switch (name) {
+        case 'open':
+          this.dispatchEvent(new Event('attrchange.open'));
+          break;
+        case 'value':
+        case 'range-start':
+        case 'range-end':
+          this.dispatchEvent(new Event('attrchange.value'));
+          break;
+      }
+    }
+  }
+);
+
+// Date functions
+
+function sameDate(a, b) {
+  return a?.getTime() === b?.getTime();
+}
+
+function toISODateString(date) {
+  // The 'en-ca' locale gives us the local date in ISO format,
+  // and it's not the same as as the actual ISO date
+  return date?.toLocaleDateString('en-ca');
+}
+
+function toOptionalDate(dateString) {
+  const date = new Date(dateString);
+  if (!date.getTime()) {
+    return;
+  }
+  date.setHours(0, 0, 0, 0);
+  return date;
+}
+
+function formatDate(formatString, date) {
+  const year = `${date.getFullYear()}`;
+  const month = `${date.getMonth() + 1}`;
+  const day = `${date.getDate()}`;
+
+  return formatString
+    .replace(/yy+/, function(s) {
+      if (s.length === 2) {
+        return year.slice(3);
+      }
+      return year;
+    })
+    .replace(/m{1,2}/, function(s) {
+      return month.padStart(s.length, '0');
+    })
+    .replace(/d{1,2}/, function(s) {
+      return day.padStart(s.length, '0');
+    });
+}
+
+// Parser
+
+function createDateParser(formatString) {
+  const tokens = [];
+  const tokenTypesFound = new Set();
+  let lastCharType = TOKEN_START;
+  const buffer = [];
+
+  formatString = formatString.toLowerCase();
+
+  // Tokenize
+  for (let i = 0; i < formatString.length; i++) {
+    const char = formatString[i];
+    const type = CHAR_TO_TOKEN_TYPE[char] || TOKEN_ANY;
+
+    if (type !== lastCharType && buffer.length) {
+      tokenTypesFound.add(lastCharType);
+      tokens.push({
+        type: lastCharType,
+        text: buffer.join(''),
+      });
+      buffer.length = 0;
+    }
+
+    buffer.push(char);
+    lastCharType = type;
+  }
+
+  if (buffer.length) {
+    tokenTypesFound.add(lastCharType);
+    tokens.push({
+      type: lastCharType,
+      text: buffer.join(''),
+    });
+  }
+
+  // Sanity check
+  if (
+    !tokenTypesFound.has(TOKEN_YEAR) ||
+    !tokenTypesFound.has(TOKEN_MONTH) ||
+    !tokenTypesFound.has(TOKEN_DATE)
+  ) {
+    throw Error('Invalid date format: date format must include year, month, and date');
+  }
+
+  // Convert tokens to regex
+  let regexSource = '^';
+  for (const token of tokens) {
+    switch (token.type) {
+      case TOKEN_YEAR:
+        regexSource += '(?<year>\\d{2}|\\d{4})';
```

Comment:
I would NOT accept `yy` here. Besides that fact that `yy` generally sucks, I don't think we can safely assume that `24.12.20` is `1920-12-24` (and not `2020-12-24`).

I'd prefer to always require `yyyy` for the year. That way people get a validation error and can fix their date.